build/lib/src/builder/build_step_impl.dart

```diff
@@ -89,14 +90,14 @@ class BuildStepImpl implements BuildStep {
   Future writeAsString(AssetId id, FutureOr<String> content,
       {Encoding encoding: UTF8}) {
     _checkOutput(id);
-    var done = _futureOrWrite(
+    var done = _futureOrWrite<String>(
         content, (c) => _writer.writeAsString(id, c, encoding: encoding));
     _outputsCompleted = _outputsCompleted.then((_) => done);
     return done;
   }
 
   Future _futureOrWrite<T>(FutureOr<T> content, Future write(T content)) =>
-      (content is Future<T>) ? content.then(write) : write(content);
+      (content is Future<T>) ? content.then(write) : write(content as T);
```

Comment:
This `as T` cast seems weird as well... I guess the type inference just isn't smart enough here but it should be able to be? Maybe worth filing a bug on the language team.